### PR TITLE
EntityHandleManager and SDK Handle Facades

### PR DESCRIPTION
**Pull request by cromwellian**
Introduce EntityHandleManager
* creates singleton/set handles given a HandleHolder (code gen), handleName, StorageKey, mode
* Backed by a core HandleManager
* Wires up facades around core Handles to provide Read/Write{Singleton|Set} versions
* Populates HandleHolder for particles to access created handles
* Delegates incoming Callbacks onUpdate event to SDK Handle onUpdate callbacks
* Moves to Entity from JvmEntity (no longer needed after WASM separation)
* Plumbs 'canRead' into core Handle Impls
** Note: need onSync/DeSync in handle interfaces as well
** Tests

Usage of this class from within an ArcHost will be something like this:

```
val particle = instantiateParticle(particleSpec.particleName)
val handleManager = entityHandleManager()
particleSpec.handles.forEach {
  handleManager.createSingletonHandle(
    particle.handles,
    it.name, 
    it.value.storageKey, 
    it.value.type.toSchema(), 
    it.value.handleMode)
}
```
After this loop, particles.handles will contain all of the handles initialized, and it is safe to invoke particle.onCreate().  It's a little more complex than that, since you need to examine the Schema to determine if you need to call createSingletonHandle() or createSetHandle()

The gist is that particles can access handles via static typed field lookups like particle.handles.fooHandle, and ArcHosts can access handles dynamically via the internal map on the HandleHolder.

### CODE REVIEW

**Comment by cromwellian** [comment]
/gcbrun

Check flaky DB test

File: java/arcs/core/host/SdkHandleManager.kt [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    }
+
+    protected suspend fun fireUpdate(value: T) {
+        onUpdateActions.forEach { action -> action(value) }
+    }
+}
+
+internal open class ReadableSingletonHandleImpl<T : Entity>(
+    val entitySpec: EntitySpec<T>,
+    val handleName: String,
```
Comment (jblebrun) [inline java/arcs/core/host/SdkHandleManager.kt]: Is there a case where we want the `handleName` of the wrapped handle to be different than that of the storage handle? If not, we could remove this property, and have the `name` property below return the `storageHandle` name.

File: java/arcs/core/host/EntitySpecNotFound.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.host
+
+/**
+ * Called when no [arcs.sdk.JvmEntitySpec] was found in the [Particle]'s associated
+ * [HandleHolder.entitySpec] map. This can happen if a [Handle] has been created for a
+ * [Particle] which had declaration for it in it's manifest and may indicate that the
```
Comment (jblebrun) [inline java/arcs/core/host/EntitySpecNotFound.kt]: Nits: it's -> its (fun to say!)

File: java/arcs/core/host/SdkHandleManager.kt [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+     *
+     * @property handleHolder contains handle and entitySpec declarations
+     * @property handleName name for the handle, must be present in [HandleHolder.entitySpecs]
+     * @property storageKey a [StorageKey]
+     * @property schema the [Schema] for this [StorageKey]
+     * @property handleMode whether a handle is Read,Write, or ReadWrite (default)
+     * @property sender block used to execute callback lambdas
+     * @property idGenerator used to generate unique IDs for newly stored entities.
+     */
+    suspend fun sdkSingletonHandle(
```
Comment (shaper) [inline java/arcs/core/host/SdkHandleManager.kt]: nit: unless this class provides ways to create both sdk and non-sdk handles I would think just `singletonHandle` or `createSingletonHandle` would be sufficient here (and similar for below methods), since `sdk` is implicit in the containing class name, and a verb prefix is often common (but I am a noob re: Kotlin idioms, so definitely defer to others).

File: java/arcs/core/storage/Handle.kt [hunk 4]
```diff
@@ -69,7 +69,7 @@ open class Handle<Data : CrdtData, Op : CrdtOperationAtTime, T>(
     val storageProxy: StorageProxy<Data, Op, T>,
 
     /** [callback] contains optional Handle-owner provided callbacks to add behavior. */
-private val callback: Callbacks<Data, Op, T>? = null,
+    var callback: Callbacks<Data, Op, T>? = null,
```
Comment (shaper) [inline java/arcs/core/storage/Handle.kt]: Why this change, is this ok long-term or an interim change, should we update docs?

File: java/arcs/sdk/BUILD [hunk 5]
```diff
@@ -2,7 +2,7 @@ load(
     "//third_party/java/arcs/build_defs:build_defs.bzl",
     "arcs_kt_jvm_library",
     "arcs_kt_native_library",
-    "kt_js_library",
+    #   "kt_js_library",
```
Comment (shaper) [inline java/arcs/sdk/BUILD]: Add a comment here as well (can point to the comment below).

File: javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        } catch (e: Exception) {
+            assertThat(e).isInstanceOf(NoSuchElementException::class.java)
+            assertThat(e.message).isEqualTo(
+                "Handle ${handleName} not initialized in TestParticle"
+            )
+        }
+    }
+
+    @Test
+    fun testCreateSingletonHandle() = runBlockingTest {
```
Comment (shaper) [inline javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt]: nit: `test` prefix usually not used for test method names when using JUnit that has `@Test` annotation, since it's implicit, same rest of PR.

File: third_party/java/arcs/build_defs/internal/kotlin.bzl [hunk 7]
```diff
@@ -58,7 +58,7 @@ DISABLED_LINT_CHECKS = [
 ALL_PLATFORMS = ["jvm", "js", "wasm"]
 
 # Default set of platforms for Kotlin libraries.
-DEFAULT_LIBRARY_PLATFORMS = ["jvm", "js"]
+DEFAULT_LIBRARY_PLATFORMS = ["jvm"]
```
Comment (shaper) [inline third_party/java/arcs/build_defs/internal/kotlin.bzl]: Probably we want a TODO here as well re: restoring for Kotlin-JS?

File: javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.coroutines.experimental.suspendCoroutine
+
+typealias Person = TestParticleInternal1
+
+@Suppress("EXPERIMENTAL_API_USAGE")
+@RunWith(AndroidJUnit4::class)
+class AndroidSdkHandleManagerTest {
```
Comment (shaper) [inline javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt]: As you'd mentioned, these tests are pretty monolithic. Perhaps it's worth capturing while it's fresh in your mind a granular list (in a comment, or in a filed issue) of the unit tests you'd write right now if you had more time, so that someone can field soon following and we don't forget or miss things.

File: src/tools/schema2kotlin.ts [hunk 9]
```diff
@@ -304,6 +306,6 @@ ${typeDecls.length ? typeDecls.join('\n') + '\n' : ''}`;
   }
 
   private getType(type: string): string {
-    return this.opts.wasm ? `Wasm${type}` : `Jvm${type}`;
+    return this.opts.wasm ? `Wasm${type}` : `${type}`;
```
Comment (shaper) [inline src/tools/schema2kotlin.ts]: I know there's been shuffling around Jvm, I don't have enough of the latest background to validate what's being done here, @csilvestrini to comment.

File: java/arcs/core/host/SdkHandleManager.kt [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+import kotlinx.coroutines.launch
+
+typealias Sender = (block: suspend () -> Unit) -> Unit
+typealias SingletonSenderCallbackAdapter<E> =
+    SenderCallbackAdapter<SingletonData<RawEntity>, SingletonOp<RawEntity>, RawEntity?, E?>
+typealias CollectionSenderCallbackAdapter<E> =
+    SenderCallbackAdapter<SetData<RawEntity>, SetOp<RawEntity>, Set<RawEntity>, E>
+
+/**
+ * Wraps a [HandleManager] and creates SDK handles based on [HandleMode], such as
```
Comment (shaper) [inline java/arcs/core/host/SdkHandleManager.kt]: Do users have to use the HandleManager? Can they use handles directly? How do these handles relate to non-SDK handles?

Could imagine elaborating on these in the doc comment here.

File: java/arcs/core/host/SdkHandleManager.kt (hunk 3 above)
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: I tried to pattern this after the HandleManager which omits the 'create' prefix, but I can change it.

File: java/arcs/core/storage/Handle.kt (hunk 4 above)
Comment (cromwellian) [reply java/arcs/core/storage/Handle.kt]: It originally was public mutable when I wrote my PR, then after I rebased, someone changed it to private immutable. I can change it back, but it means making a field in the Adapter class mutable, because StorageHandles have to be created before their SdkHandle wrapper, but the Callbacks adapter needs to reference SdkHandle in its ctor. There's no topological sort that works. :)

I can add a SetHandle() to the SenderCallbacksAdapter and modify it afterwards.

File: javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt (hunk 8 above)
Comment (cromwellian) [reply javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt]: We can port the SIngletonImplTest/CollectionImpltests that were deleted. They are micro tests of basic functionality. 

Not tested here is clear(), remove()

File: src/tools/schema2kotlin.ts (hunk 9 above)
Comment (cromwellian) [reply src/tools/schema2kotlin.ts]: Jvm* types were added because of WASM. I hoped the the serialize/deserialize entitySpec stuff up to EntitySpec/Entity. 

The point here is that we don't want core to depend on SDK. These Handle classes are in Core, and need to invoke serialize/deserialize, ergo, they need to be in core. Which is now safe to do because of Cameron's changes.

File: java/arcs/core/host/SdkHandleManager.kt (hunk 10 above)
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: Yes, you have to use HandleManager. There are private methods we could make public that take Storage handles directly, we could expose them later, but right now, even those users who don't go through Particle API to access handles are expected to use HandleManager.

Also, HandleManager abstracts away Core vs Android, so it provides additional benefit.

File: java/arcs/core/host/SdkHandleManager.kt (hunk 1 above)
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: I'm unsure of this, Yuan reported that if two storage handles have the same name, it crashes storage service. Right now the StorageHandles use StorageKey.toString() as the name. I would say in the interests of expediency, because this code is working and not crashing, we keep it like this, and then have a meeting with Jason and others over the naming issue to get a good idea on what the Storage layer wants.

File: java/arcs/core/host/SdkHandleManager.kt (hunk 3 above)
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: Done

File: java/arcs/core/storage/Handle.kt (hunk 4 above)
Comment (cromwellian) [reply java/arcs/core/storage/Handle.kt]: I think we can switch this in a followup.

File: javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+                handleHolder.writeSetHandle
+            }
+
+            val writeSetHandle = hm.createSetHandle(
+                handleHolder,
+                writeSetHandleName,
+                setKey,
+                schema,
+                HandleMode.Write
+            )
```
Comment (csilvestrini) [inline javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt]: Optional: might be more readable to create a helper method for these. It looks like the arguments rarely change.

```
private fun createSetHandle(
    handleManager: HandleManager,
    handleName: String,
    handleMode: HandleMode
) = handleManager.createSetHandle(
    handleHolder,
    handleName,
    setKey,
    schema,
    handleMode
)
```

(`handleHolder` could be a member variable and be initialised in `setUp`)

File: java/arcs/core/host/EntitySpecNotFound.kt [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+ * manifest and generated code are out of sync.
+ */
+class EntitySpecNotFound(
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply java/arcs/core/host/EntitySpecNotFound.kt]: Deleted

File: java/arcs/core/host/SdkHandleManager.kt [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+ * [ReadableSingleton] for [HandleMode.Read]. To obtain a [HandleHolder], use
+ * `arcs_kt_schema` on a manifest file to generate a `{ParticleName}Handles' class, and
+ * invoke its default constructor, or obtain it from the [BaseParticle.handles] field.
+ */
+class SdkHandleManager(val handleManager: HandleManager) {
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: Renamed to EntityHandleManager

File: java/arcs/core/host/SdkHandleManager.kt [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+
+    override fun onSync(handle: arcs.core.storage.Handle<Data, Op, T>) {
+    }
+
+    override fun onDesync(handle: arcs.core.storage.Handle<Data, Op, T>) {
+    }
+}
+
+internal fun <T : Entity> RawEntity.deserialize(entitySpec: EntitySpec<T>) =
+    entitySpec.deserialize(this)
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: I did the former.  I'm hesitate to add SDK notions like entity serialization/deserialization to raw storage layer stuff until talking it through with Jason/Shane, I'm trying to keep the SDK stuff fairly layered on top.

File: java/arcs/core/host/SdkHandleManager.kt [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+     */
+    private fun invokeWithSender(block: suspend () -> Unit) = sender(block)
+
+    override fun onUpdate(handle: arcs.core.storage.Handle<Data, Op, T>, op: Op) {
+        invokeWithSender {
+            updateCallback.invoke(fetchFunc())
+        }
+    }
+
+    override fun onSync(handle: arcs.core.storage.Handle<Data, Op, T>) {
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply java/arcs/core/host/SdkHandleManager.kt]: I added implementations for these, no longer empty

File: java/arcs/core/storage/handle/HandleManager.kt [hunk 16]
```diff
@@ -110,7 +112,7 @@ class HandleManager(private val aff: ActivationFactoryFactory? = null) {
             }
         }
 
-        return SetHandle(name, storageProxy, callbacks, refinement).also {
+        return SetHandle(name, storageProxy, callbacks, refinement, canRead).also {
```
Comment (Cypher1) [inline java/arcs/core/storage/handle/HandleManager.kt]: Possibly switch refinement and canRead around?

File: javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt (hunk 11 above)
Comment (cromwellian) [reply javatests/arcs/android/host/AndroidSdkHandleManagerTest.kt]: Done

File: java/arcs/core/host/HandleHolder.kt [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Interface used by [ArcHost]s to interact dynamically with code-generated [Handle] fields
+ * used by [Particle]s.
+ *
+ * @property map Key is a handle name, value is the corresponding [Handle].
+ * @property entitySpecs Key is a handle name, value is the corresponding [EntitySpec].
+ */
+interface HandleHolder {
+    val map: Map<String, Handle>
```
[the remark this answers is not in the document]
Comment (cromwellian) [reply java/arcs/core/host/HandleHolder.kt]: Done

File: java/arcs/core/storage/handle/HandleManager.kt (hunk 16 above)
Comment (cromwellian) [reply java/arcs/core/storage/handle/HandleManager.kt]: I kept it this way in case someone called this from somewhere else without using named parameters.

File: java/arcs/core/storage/api/Entity.kt [hunk 18]
```diff
@@ -24,4 +27,11 @@ interface Entity {
 interface EntitySpec<T : Entity> {
     /** Returns an empty new instance of [T]. */
     fun create(): T
+
+    /**
+     * Takes a [RawEntity] representing a serialized representation of a [RawEntity] from the
```
Comment (csilvestrini) [inline java/arcs/core/storage/api/Entity.kt]: I think there's still some redundant stuff here. Just: `Takes a [RawEntity] and convert it to concrete entity class [T].`

File: javatests/arcs/android/host/AndroidEntityHandleManagerTest.kt [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+
+    val entity1 = Person("Jason", 21.0, false)
+    val entity2 = Person("Jason", 22.0, true)
+    lateinit var handleHolder: TestParticleHandles
+
+    private val schema = Schema(
+        listOf(SchemaName("Person")), SchemaFields(
+            singletons = mapOf(
+                "name" to FieldType.Text, "age" to FieldType.Number, "is_cool" to FieldType.Boolean
+            ), collections = emptyMap()
```
Comment (csilvestrini) [inline javatests/arcs/android/host/AndroidEntityHandleManagerTest.kt]: Separate lines for args:

```
            singletons = mapOf(
                "name" to FieldType.Text,
                "age" to FieldType.Number,
                "is_cool" to FieldType.Boolean
            ),
            collections = emptyMap()
        ),
        "1234acf"
    )
```

File: javatests/arcs/android/host/AndroidEntityHandleManagerTest.kt [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+        app.setTheme(R.style.Theme_AppCompat);
+
+        // Initialize WorkManager for instrumentation tests.
+        WorkManagerTestInitHelper.initializeTestWorkManager(app)
+
+        handleHolder = TestParticleHandles()
+    }
+
+    fun handleManagerTest(block: suspend TestCoroutineScope.(EntityHandleManager) -> Unit) =
+        runBlockingTest {
```
Comment (csilvestrini) [inline javatests/arcs/android/host/AndroidEntityHandleManagerTest.kt]: If you put the arg on its own line, then the `=` and the `runBlockingTest` can go on the same line, and everything can be indented less:

```
fun handleManagerTest(
    block: suspend TestCoroutineScope.(EntityHandleManager) -> Unit
) = runBlockingTest {
    ...
}
```

File: java/arcs/core/host/EntityHandleManager.kt [hunk 21]
```diff
@@ -0,0 +1,452 @@
...
+            )
+            is SetHandle<*> -> createSetHandle(
+                entitySpec,
+                handleName,
+                storageHandle as SetHandle<RawEntity>,
+                handleMode,
+                idGenerator,
+                sender
+            )
+            else -> throw Exception("Unknown storage handle type ${storageHandle::class}")
```
Comment (jasonwyatt) [inline java/arcs/core/host/EntityHandleManager.kt]: Please don't throw base Exceptions in the future.

Comment (Cypher1) [reply java/arcs/core/host/EntityHandleManager.kt]: @jasonwyatt Got any advice on which exception class to use when? Kotlin has a different set to the ones I'm used to.
